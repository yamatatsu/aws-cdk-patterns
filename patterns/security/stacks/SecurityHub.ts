import * as cdk from "@aws-cdk/core"
import * as iam from "@aws-cdk/aws-iam"
import * as securityhub from "@aws-cdk/aws-securityhub"
import * as events from "@aws-cdk/aws-events"
import * as targets from "@aws-cdk/aws-events-targets"
import * as sns from "@aws-cdk/aws-sns"

export class SecurityHub extends cdk.Stack {
  public readonly topic: sns.ITopic

  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    new iam.CfnServiceLinkedRole(this, "ServiceLinkedRoleForSecurityHub", {
      awsServiceName: "securityhub.amazonaws.com",
      description:
        "A service-linked role required for AWS Security Hub to access your resources.",
    })

    new securityhub.CfnHub(this, "SecurityHub")

    const topic = new sns.Topic(this, "SecurityHubTopic")

    const rule = new events.Rule(this, "SecurityHubNotificationRule", {
      ruleName: "AlertSecurityHubFindings",
      description: "Alert to SNS topic when find threats by SecurityHub",
      eventPattern: {
        source: ["aws.securityhub"],
        detailType: ["Security Hub Findings - Imported"],
        detail: {
          // PASSED(問題なし)以外を1度だけ報告 参考:https://dev.classmethod.jp/articles/security-hub-filtering-examples/#toc-4
          findings: {
            Compliance: { Status: [{ "anything-but": "PASSED" }] },
            RecordState: ["ARCHIVED"],
          },
        },
      },
    })
    rule.addTarget(new targets.SnsTopic(topic))

    this.topic = topic
  }
}
