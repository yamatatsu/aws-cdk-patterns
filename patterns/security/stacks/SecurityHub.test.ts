import { SynthUtils } from "@aws-cdk/assert"
import * as cdk from "@aws-cdk/core"
import { SecurityHub } from "./SecurityHub"

test("snapshot test", () => {
  const app = new cdk.App()

  const target = new SecurityHub(app, "Target")

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot()
})
