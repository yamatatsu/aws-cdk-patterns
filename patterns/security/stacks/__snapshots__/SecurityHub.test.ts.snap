// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "SecurityHub": Object {
      "Type": "AWS::SecurityHub::Hub",
    },
    "SecurityHubNotificationRule02379247": Object {
      "Properties": Object {
        "Description": "Alert to SNS topic when find threats by SecurityHub",
        "EventPattern": Object {
          "detail": Object {
            "findings": Object {
              "Compliance": Object {
                "Status": Array [
                  Object {
                    "anything-but": "PASSED",
                  },
                ],
              },
              "RecordState": Array [
                "ARCHIVED",
              ],
            },
          },
          "detail-type": Array [
            "Security Hub Findings - Imported",
          ],
          "source": Array [
            "aws.securityhub",
          ],
        },
        "Name": "AlertSecurityHubFindings",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "SecurityHubTopicA93309F7",
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "SecurityHubTopicA93309F7": Object {
      "Type": "AWS::SNS::Topic",
    },
    "SecurityHubTopicPolicyE02112B5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
              "Resource": Object {
                "Ref": "SecurityHubTopicA93309F7",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "SecurityHubTopicA93309F7",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "ServiceLinkedRoleForSecurityHub": Object {
      "Properties": Object {
        "AWSServiceName": "securityhub.amazonaws.com",
        "Description": "A service-linked role required for AWS Security Hub to access your resources.",
      },
      "Type": "AWS::IAM::ServiceLinkedRole",
    },
  },
}
`;
